### PR TITLE
Threaded ctl command runner

### PR DESCRIPTION
**Pull request by lbudai**
Split #3349 

First part of the PR is:
 * run stats/query commands in a separate thread

### CODE REVIEW

**Comment by kira-syslogng** [comment]
Build FAILURE

**Comment by kira-syslogng** [comment]
Build SUCCESS

**Comment by kira-syslogng** [comment]
Build SUCCESS

File: lib/control/control-server.c [hunk 1]
```diff
@@ -41,11 +166,27 @@ void
...
 }
 
 void
 control_server_free(ControlServer *self)
 {
+  // it's not racy as the iv_main() is executed by this thread
+  // posted events are not executed as iv_quit() is already called (before control_server_free)
+  // but it is possible that some ThreadedCommandRunner::thread are still running
+  if (self->worker_threads)
+    g_list_free_full(self->worker_threads, _delete_thread_command_runner);
```
Comment (furiel) [inline lib/control/control-server.c:189]: Unfortunately this does not cancel the long running commands. `control_server_free` is called too late: after main loop is not running. For example, syslog-ng-ctl stats is running in mainloop, so main_loop_deinit will not be called until it is finished.

You can try this:

```
@version: 3.29

options {
 stats-level(3);
};

log {
  source { example-msg-generator(freq(0) num(100000000)
                                 values("HOST" => "host$(iterate $(+ 1 $_) 0)") keep-hostname(yes)); };
  destination { file(/dev/null); };
};
```

Let syslog-ng to run a bit: example-msg-generator will generate a bit slowly, the longer you wait, the longer stats runs. After waiting a bit, call syslog-ng-ctl stats and syslog-ng-ctl stop at the same time. syslog-ng-ctl-stop will "hang".

Comment (lbudai) [reply lib/control/control-server.c:189]: @furiel : Thanks!!
(It works perfectly with my original branch, when control server is run in a different thread - I'll fix it soon).

**Comment by kira-syslogng** [comment]
Build SUCCESS

File: lib/control/control-server.c [hunk 2]
```diff
@@ -26,10 +26,181 @@
...
+  ThreadedCommandRunner *runner = _thread_command_runner_new(self, command, user_data);
+  _thread_command_runner_run(runner, cmd_cb);
+}
+
+static void
+_delete_thread_command_runner(gpointer data)
+{
+  ThreadedCommandRunner *self = (ThreadedCommandRunner *) data;
+  g_assert(self->real_thread.tid_saved == TRUE);
+  gboolean has_to_free = FALSE;
```
Comment (szemere) [inline lib/control/control-server.c:175]: discussed IRL: replaced by https://github.com/syslog-ng/syslog-ng/pull/3387/files#diff-c8826ab0571c7b2f08d7f66e83e5c29aR125

File: lib/control/control-server.c [hunk 3]
```diff
@@ -26,10 +26,181 @@
...
+      _thread_command_runner_sync_run(self, func);
+      return;
+    }
+
+  iv_event_register(&self->response_received);
+
+  self->thread = g_thread_new(self->command->str, (GThreadFunc) _thread, self);
+  _thread_command_runner_wait_for_tid_saved(self);
+  ControlServer *server = self->connection->server;
+  server->worker_threads = g_list_append(server->worker_threads, self);
```
Comment (szemere) [inline lib/control/control-server.c:159]: nitpicking: g_list_prepend would give a better performance

**Comment by kira-syslogng** [comment]
Build SUCCESS

File: lib/control/control-server.c (hunk 1 above)
Comment (lbudai) [reply lib/control/control-server.c:228]: Unfortunately only my original solution works... the reason is very simple: `stats_lock`...
When generator source (or any other source executed by the main thread) produces messages, it also updates its stat counters.  When the stats mutex is locked by a `syslog-ng-ctl stats` then ivykis will wait the actual 'job' to be finished (which can happen only after the mutex is unlocked) - until that it won't process the events of the control loop...
This is why we need to execute `ControlServer` in its own thread: that won't lock the stats mutex.

**Review by furiel — APPROVED** [review]

**Review by szemere — APPROVED** [review]

**Pull request merged**